File: backend/app/routers/ingredients.py
from typing import Literal
from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel

from app.models import Ingredient, IngredientSession
from app.services.ingredient_parser import IngredientParser
from app.services.session_store import session_store

router = APIRouter(prefix="/api/ingredients", tags=["ingredients"])

# Initialize parser
parser = IngredientParser()


# Request/Response models
class ParseRequest(BaseModel):
    text: str


class ParseResponse(BaseModel):
    ingredients: list[Ingredient]


class CreateSessionRequest(BaseModel):
    user_id: str


class UpdateStatusRequest(BaseModel):
    status: Literal["in_progress", "confirmed", "used"]


# Endpoints
@router.post("/parse", response_model=ParseResponse)
async def parse_ingredients(request: ParseRequest):
    """Parse natural language text into structured ingredients."""
    ingredients = await parser.parse(request.text)
    return ParseResponse(ingredients=ingredients)


@router.post("/sessions", response_model=IngredientSession, status_code=201)
async def create_session(request: CreateSessionRequest):
    """Create a new ingredient collection session."""
    session = session_store.create_session(request.user_id)
    return session


@router.get("/sessions/latest/{user_id}", response_model=IngredientSession)
async def get_latest_session(user_id: str):
    """Get the most recent session for a user."""
    session = session_store.get_latest_session(user_id)
    if not session:
        raise HTTPException(status_code=404, detail="No session found for user")
    return session


@router.get("/sessions/{session_id}", response_model=IngredientSession)
async def get_session(session_id: str):
    """Get a specific session by ID."""
    session = session_store.get_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return session


@router.post("/sessions/{session_id}/ingredients", response_model=IngredientSession)
async def add_ingredients(session_id: str, request: ParseResponse):
    """Add ingredients to a session."""
    session = session_store.add_ingredients(session_id, request.ingredients)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return session


@router.delete("/sessions/{session_id}/ingredients/{ingredient_id}", response_model=IngredientSession)
async def remove_ingredient(session_id: str, ingredient_id: str):
    """Remove an ingredient from a session."""
    session = session_store.remove_ingredient(session_id, ingredient_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return session


@router.patch("/sessions/{session_id}/status", response_model=IngredientSession)
async def update_status(session_id: str, request: UpdateStatusRequest):
    """Update the status of a session."""
    session = session_store.update_status(session_id, request.status)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return session

File: frontend/src/api/ingredients.ts
import { apiClient } from './client';
import type { ParseResponse, IngredientSession, Ingredient } from '../types';

/**
 * API client for ingredient-related endpoints.
 * Provides methods for parsing text, managing sessions, and manipulating ingredients.
 */

export const ingredientsApi = {
  /**
   * Parse natural language text into structured ingredients.
   * @param text - Raw text describing ingredients (e.g., "2 apples, 1 lb chicken")
   * @returns Parsed ingredients with quantities and units
   */
  async parse(text: string): Promise<ParseResponse> {
    const response = await apiClient.post<ParseResponse>('/api/ingredients/parse', { text });
    return response.data;
  },

  /**
   * Create a new ingredient collection session for a user.
   * @param userId - User identifier
   * @returns New session object
   */
  async createSession(userId: string): Promise<IngredientSession> {
    const response = await apiClient.post<IngredientSession>('/api/ingredients/sessions', {
      user_id: userId,
    });
    return response.data;
  },

  /**
   * Get the latest ingredient session for a user.
   * @param userId - User identifier
   * @returns Latest session or null if no session exists
   */
  async getLatestSession(userId: string): Promise<IngredientSession | null> {
    try {
      const response = await apiClient.get<IngredientSession>(
        `/api/ingredients/sessions/latest/${userId}`
      );
      return response.data;
    } catch (error: any) {
      if (error.response?.status === 404) {
        return null;
      }
      throw error;
    }
  },

  /**
   * Get a specific ingredient session by ID.
   * @param sessionId - Session identifier
   * @returns Session object
   */
  async getSession(sessionId: string): Promise<IngredientSession> {
    const response = await apiClient.get<IngredientSession>(
      `/api/ingredients/sessions/${sessionId}`
    );
    return response.data;
  },

  /**
   * Add ingredients to an existing session.
   * @param sessionId - Session identifier
   * @param ingredients - Array of ingredients to add
   * @returns Updated session object
   */
  async addIngredients(
    sessionId: string,
    ingredients: Ingredient[]
  ): Promise<IngredientSession> {
    const response = await apiClient.post<IngredientSession>(
      `/api/ingredients/sessions/${sessionId}/ingredients`,
      { ingredients }
    );
    return response.data;
  },

  /**
   * Remove a specific ingredient from a session.
   * @param sessionId - Session identifier
   * @param ingredientId - Ingredient identifier to remove
   * @returns Updated session object
   */
  async removeIngredient(sessionId: string, ingredientId: string): Promise<IngredientSession> {
    const response = await apiClient.delete<IngredientSession>(
      `/api/ingredients/sessions/${sessionId}/ingredients/${ingredientId}`
    );
    return response.data;
  },

  /**
   * Update the status of an ingredient session.
   * @param sessionId - Session identifier
   * @param status - New status ('in_progress' | 'confirmed' | 'used')
   * @returns Updated session object
   */
  async updateStatus(
    sessionId: string,
    status: 'in_progress' | 'confirmed' | 'used'
  ): Promise<IngredientSession> {
    const response = await apiClient.patch<IngredientSession>(
      `/api/ingredients/sessions/${sessionId}/status`,
      { status }
    );
    return response.data;
  },
};
